Sming/system/include/debug_progmem.h

```diff
@@ -52,23 +57,39 @@ extern "C" {
 #define MACROQUOT(x) #x
 #define MACROQUOTE(x) MACROQUOT(x)
 
+#ifdef SMING_INCLUDED
+#define PROGMEM_DEBUG                                                                                                  \
+	__attribute__((aligned(4))) __attribute__((section(MACROQUOTE(MACROCONCAT(.irom.debug., __COUNTER__, __LINE__)))))
+#else
+#define PROGMEM_DEBUG
+#endif
+
 //A static const char[] is defined having a unique name (log_ prefix, filename and line number)
 //This will be stored in the irom section(on flash) feeing up the RAM
 //Next special version of printf from FakePgmSpace is called to fetch and print the message
 #if DEBUG_PRINT_FILENAME_AND_LINE
-#define debug_e(fmt, ...) \
-	({static const char log_string[] \
-	__attribute__((aligned(4))) \
-	__attribute__((section(MACROQUOTE(MACROCONCAT(.irom.debug.,__COUNTER__,__LINE__))))) = "[" MACROQUOTE(CUST_FILE_BASE) ":%d] " fmt "\n"; \
-	printf_P_stack(log_string, __LINE__, ##__VA_ARGS__);})
+#define debug_e(fmt, ...)                                                                                              \
+	(__extension__({                                                                                                   \
+		static const char log_string[] PROGMEM_DEBUG = "[" MACROQUOTE(CUST_FILE_BASE) ":%d] " fmt "\n";                \
+		printf_P(log_string, __LINE__, ##__VA_ARGS__);                                                                 \
+	}))
 #else
-#define debug_e(fmt, ...) \
-	({static const char log_string[] \
-	__attribute__((aligned(4))) \
-	__attribute__((section(MACROQUOTE(MACROCONCAT(.irom.debug.,__COUNTER__,__LINE__))))) = fmt "\n"; \
-	printf_P_stack(log_string, ##__VA_ARGS__);})
+#define debug_e(fmt, ...)                                                                                              \
+	(__extension__({                                                                                                   \
+		static const char log_string[] PROGMEM_DEBUG = "%u " fmt "\n";                                                 \
```

Comment:
There might be complaints that we prepend the time to the message. If that happens we should add directive that controls this. For now no action from your side is required.

Reply:
Something different: with the latest changes my app started to crash. Stacktrace shows the following as a reason for it:

```
python $SMING_HOME/../tools/decode-stacktrace.py out/build/app_0.out  crash.log 

0x4027d72e: std::_Function_handler<unsigned int (char const*, unsigned int), std::_Bind<unsigned int (*(uart_*, std::_Placeholder<1>, std::_Placeholder<2>))(uart_*, void const*, unsigned int)> >::_M_invoke(std::_Any_data const&, char const*, unsigned int) at /path/to/Espressif/xtensa-lx106-elf/xtensa-lx106-elf/include/c++/4.8.5/functional:2058
0x40263852: std::function<unsigned int (char const*, unsigned int)>::operator()(char const*, unsigned int) const at /path/to/dev/Sming/Sming/system/m_printf.cpp:277
0x402638a3: m_vprintf at /path/to/dev/Sming/Sming/system/m_printf.cpp:277
```

Any idea?

Reply:
Hm.. a debug_e statement started crashing my app:
```
	debug_d("Mark size: %d", mark.size);
```

Have to investigate a bit further to narrow it down.

Reply:
What's the exception code?

Reply:
11441506 Crashes for the current ROM(0): 21
11487320 ... after 10058 ms
11516511 Reason for last exception: 3 ( see: https://arduino-esp8266.readthedocs.io/en/latest/exception_causes.html)

I created a sample code that reproduces the issue. Make a new branch for the test

```
cd %SMING_HOME%
git checkout develop
git pull
git branch tmp
git checkout tmp
make dist-clean
```

and then apply the patch below:
```diff
diff --git a/samples/Basic_rBoot/Makefile-user.mk b/samples/Basic_rBoot/Makefile-user.mk
index fcc217a..726ea7b 100644
--- a/samples/Basic_rBoot/Makefile-user.mk
+++ b/samples/Basic_rBoot/Makefile-user.mk
@@ -40,6 +40,7 @@ SPIFF_FILES = files
 #### overridable rBoot options ####
 ## use rboot build mode
 RBOOT_ENABLED ?= 1
+RBOOT_SILENT = 1
 ## enable big flash support (for multiple roms, each in separate 1mb block of flash)
 RBOOT_BIG_FLASH ?= 1
 ## two rom mode (where two roms sit in the same 1mb block of flash)
@@ -64,3 +65,5 @@ SPIFF_SIZE      ?= 65536
 ## esptool2 path
 #ESPTOOL2        ?= esptool2
 
+
+DEBUG_VERBOSE_LEVEL=3
diff --git a/samples/Basic_rBoot/app/application.cpp b/samples/Basic_rBoot/app/application.cpp
index 3fb5642..65ad65d 100644
--- a/samples/Basic_rBoot/app/application.cpp
+++ b/samples/Basic_rBoot/app/application.cpp
@@ -159,6 +159,27 @@ void serialCallBack(Stream& stream, char arrivedChar, unsigned short availableCh
                        }
                } else if(!strcmp(str, "info")) {
                        ShowInfo();
+               } else if(!strcmp(str, "test")) {
+                       typedef struct {
+                               uint32_t size = 0;
+                               uint32_t minVersion = 0;
+                               uint32_t checksum = 0;
+                               uint32_t reserved;
+                       } FwUpdateMark;
+
+                       FwUpdateMark mark;
+
+                       uint32_t startAddr = 0x301000;
+
+                       int markSize = sizeof(FwUpdateMark);
+
+                       uint8_t *addr = (uint8_t *)&mark;
+                       noInterrupts();
+                       spi_flash_read(startAddr, (uint32_t*)((void*)addr), markSize);
+                       interrupts();
+
+                       debug_d("Mark size: %d", mark.size);
+
                } else if(!strcmp(str, "help")) {
                        Serial.println();
                        Serial.println("available commands:");
@@ -173,6 +194,7 @@ void serialCallBack(Stream& stream, char arrivedChar, unsigned short availableCh
                        Serial.println("  ls - list files in spiffs");
                        Serial.println("  cat - show first file in spiffs");
 #endif
+                       Serial.println("  test - tests a code that crashes in my app");
                        Serial.println();
                } else {
                        Serial.println("unknown command");
```

Once ready with the patching go to Basic_rBoot
```
cd ../Basic_rBoot
make
```
Then flash the code to your device. Now type
help // just to see that the command processor is still working
test // to execute the problematic code
try to type `help` again. At least for me the device stall and restarted with software watchdog reset.

Reply:
I tried the code as-is, and it worked without any issues. Have now rebuilt against current develop and I'm getting `system_partition_table_regist: failed` - any ideas?

Reply:
`system_partition_table_regist` - SDK didn't auto-update, sorted.

Reply:
> I'm getting system_partition_table_regist: failed - any ideas?

make dist-clean

Reply:
Done that. Deleted third-party, complete re-build, `make flashinit`. I cannot see any reason why `system_partition_table_regist` would fail. I added some debug output to `user_pre_init()` and values all look fine. So I just ignore the error return and continue. Works fine.

```
 ets Jan  8 2013,rst cause:2, boot mode:(3,6)

load 0x40100000, len 1340, room 16 
tail 12
chksum 0xe9
ho 0 tail 12 room 4
load 0x3ffe8000, len 660, room 12 
tail 8
chksum 0xc8
csum 0xc8

rBoot v1.4.2 - richardaburton@gmail.com
Flash Size:   32 Mbit
Flash Mode:   DIO
Flash Speed:  80 MHz
rBoot Option: Big flash

Booting rom 0.
p 101 size 0xfdff0 ,0x2000 err
system_partition_table_regist: failed
size_map = 4
partition[0]: 1, 0x00000000, 0x00001000
partition[1]: 4, 0x003fb000, 0x00001000
partition[2]: 5, 0x003fc000, 0x00001000
partition[3]: 6, 0x003fd000, 0x00003000
partition[4]: 100, 0x00001000, 0x00001000
partition[5]: 101, 0x00002000, 0x000fdff0
>> 0RRQrf cal sector: 1019
freq trace enable 0
rf[112] 119215 trying to mount spiffs at 0x00100000, length 65536
119499 fs.start: size:64 Kb, offset:0x100000

121012 mount res: 0


Currently running rom 0.
Type 'help' and press enter for instructions.

mode : null
help

available commands:
  help - display this message
  ip - show current ip address
  connect - connect to wifi
  restart - restart the esp8266
  switch - switch to the other rom and reboot
  ota - perform ota update, switch rom and reboot
  info - show esp8266 info
  ls - list files in spiffs
  cat - show first file in spiffs
  test - tests a code that crashes in my app

test
7646463 Mark size: -1
help

available commands:
  help - display this message
  ip - show current ip address
  connect - connect to wifi
  restart - restart the esp8266
  switch - switch to the other rom and reboot
  ota - perform ota update, switch rom and reboot
  info - show esp8266 info
  ls - list files in spiffs
  cat - show first file in spiffs
  test - tests a code that crashes in my app

info

SDK: v3.0.0(e27cf60)
Free Heap: 47960
CPU Frequency: 80 MHz
System Chip ID: 1c2666
SPI Flash ID: 1640c8
15086440 Count: 2
15086550 ROM 0: 8192
15088998 ROM 1: 2105344
15092202 ROM 2: 0
15094608 GPIO ROM: 0
```

Reply:
> Could it be faulty flash?

Possible, but not very likely. Because I can use the mark.size in comparison or in calculations and it's the correct value. Just debug printing causes me grief.

 > (Also, not sure why you're disabling interrupts during the flash read; not necessary, is it?)

I would say disabling interrupts is not necessary. But I was trying hard to eliminate any possible problem due to interrupts.

>  I change the 0xfdfff0 to 0xfe000 and all is well.

That is good to know. That can be updated later. 

Are you using always SDK 3.0? Or SDK 1.5 or 2.0? If the answer is lower than 3 then can you try to compile the example with the lower version SDK and try again. Just unset the SDK_BASE env variable.

Reply:
OK, so it builds and runs fine on the older SDK (version 2.0-ish I guess). It's the one included with Cherts 2.2.2 unofficial Windows SDK.

Reply:
Rebuilt under linux (virtualbox / Ubunti) and it also works fine...